rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -0,0 +1,346 @@
+        verbs: ["get", "list", "create"]
+      - resources: ["deployments"]
+        apiGroups: ["app"]
+        verbs: ["*"]
+      - resources: ["secrets"]
+        apiGroups: [""]
+        resourceNames: ["secret1", "secret2"]
+        verbs: ["get", "list""]
+```
+
+Presence of this field will signal that user wants Teleport to auto-provision these permissions into the Kubernetes clusters.
+Teleport will translate permissions defined in that field into Kubernetes RBAC definitions and will automatically create roles and bindings
+on the target Kubernetes clusters (as specified by the labels). If `namespaces` field includes '*' Teleport will create ClusterRole/ClusterRoleBinding
+pair on the Kubernetes cluster, otherwise Role/Binding pair will be created in the each listed namespace. 
+
+Subject for the bindings will be a group named `teleport:*RoleName*`, e.g. for a Teleport role `kube-role` the resulting 
+group name will be `teleport:kube-role`. This will allow Kube service to setup correct impersonation group headers based on the roles available to the user.
+
+If `kubernetes_permissions` field is set in the `allow` part of the Teleport role, fields `kubernetes_resources`, `kubernetes_users` and `kubernetes_groups` 
+should not be set, we will return an error if user will try to create a role mixing those fields. In the future we might relax this rule.
+
+The field `kubernetes_permissions` will be forbidden from appearing in the `deny` part of the role definition since we are trying to keep its meaning close to
+native Kubernetes RBAC, which can only allow access. Although the field `kubernetes_resource` could appear in the `deny` part of the role, 
+even if `kubernetes_permissions` is set in the `allow` part, giving user more flexibility using Teleport-level permissions check.
+
+Pattern matching for resource names that is supported by the `kubernetes_resources` field will not be supported (same as in native Kubernetes RBAC)
+by the `kubernetes_permissions` in the
+first implementation, but we will be able to add it later. Not allowing pattern matching, though limiting functionality compared to the `kubernetes_resources` field,
+will allow us to translate permissions defined in Teleport 1-to-1 into Kubernetes RBAC definitions. Since this feature is targeted primarily for the new users of 
+Teleport and/or users with less complicated RBAC setups, this should not be a critical issue. If we add pattern matching later, it will need to rely on 
+additional checks at the Teleport level - the way it currently works for the `kubernetes_resources`.
+
+Kubernetes RBAC will be auto-provisioned using the KubeProvisions functionality described in the next section of this RFD. For the best UX any change to
+a role that contains `kubernetes_permissions` will trigger auto-provisioning immediately to reduce latency between the moments when role was created 
+or changed and when clients can use Kubernetes cluster with the updated RBAC.
```

Comment:
Could you please clarify the reasoning behind this sentence? What is preventing us from having KubeProvisions auto-provision immediately? Can you expand on this?

A watcher in kube agents would do exactly what is needed and allow real-time synchronization in both directions, ensuring that if someone changes the cluster role in kube, the Teleport version is re-enforced and re-enforce the old version when teleport object changes.

I also find it cumbersome to have two fields defined in completely different objects that operate on the same feature and objects. From my perspective, we should have either `kubernetes_permissions` or `KubeProvisions`, but not both. People may not realize they are the same thing and could make mistakes.

Reply:
Originally KubeProvisions were just about provisioning, so it made sense with the simple approach of a reconciliation loop that we often use. Introduction of roles requires short latency for better UX, that why I added this mentioning of immediate reaction, we also trigger that on any changes to KubeProvisions. But yeah, you are right, we can add watchers for changes on the Kubernetes cluster side and reapply provisions immediately, having real-time enforcement in both directions.

> I also find it cumbersome to have two fields defined in completely different objects that operate on the same feature and objects. From my perspective, we should have either kubernetes_permissions or KubeProvisions, but not both. People may not realize they are the same thing and could make mistakes.

I think they are not really the same. And I think it's fine if users logically separate them. Maybe I should remove mentioning of `ephemeral KubeProvisions` from the RFD, seems it misleadingly shows them as something of a same. `kubernetes_provisions` logically is a different feature, it can be done through mostly the same code KubeProvisions use, but it's just an implementation details.